source/rpt/rpt_cell_reconstruction.cc

```diff
@@ -0,0 +1,626 @@
+        status = "cost_function";
+      else
+        status += "|cost_function";
+    }
+  else
+    best_cell = cells_indexes[0];
+
+  // Find the volume and the center position of the best cell and modify
+  // status when analyse positions is enable
+  for (const auto &cell : triangulation.cell_iterators_on_level(level))
+    {
+      if (cell->index() == best_cell)
+        {
+          volume                  = cell->measure();
+          reconstruction_position = cell->center(true);
+
+          // Verify if the best cell is the good cell and update status
+          if (parameters.reconstruction_param.analyse_positions)
+            {
+              if (cell->point_inside(known_positions[id]))
+                {
+                  if (status.empty())
+                    status = "right_cell";
+                  else
+                    status += "|right_cell";
+                }
+              else
+                {
+                  if (status.empty())
+                    status = "wrong_cell";
+                  else
+                    status += "|wrong_cell";
+                }
+            }
+        }
+    }
+
+  // Store information prior data exportation
+  reconstruction_positions.push_back(reconstruction_position);
+  cells_volumes.push_back(volume);
+  final_cell_level.push_back(level);
+  cell_status.push_back(status);
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::create_grid()
+{
+  TimerOutput::Scope t(computing_timer, "grid_creation");
+
+  // Generate cylinder (needs rotation and shift to get origin at the bottom
+  // with z towards top)
+  GridGenerator::subdivided_cylinder(triangulation,
+                                     5,
+                                     parameters.rpt_param.reactor_radius,
+                                     parameters.rpt_param.reactor_height / 2.);
+  GridTools::rotate(M_PI_2, 1, triangulation);
+  const Tensor<1, dim> shift_vector(
+    {0, 0, parameters.rpt_param.reactor_height / 2.});
+  GridTools::shift(shift_vector, triangulation);
+
+  // Add cylindrical manifold
+  const Tensor<1, dim>                direction({0, 0, 1});
+  const CylindricalManifold<dim, dim> manifold(direction, {0, 0, 1});
+  triangulation.set_manifold(0, manifold);
+
+  // Refine all cells
+  triangulation.prepare_coarsening_and_refinement();
+  triangulation.refine_global(
+    parameters.reconstruction_param.reactor_refinement);
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::find_vertices_positions(
+  unsigned int     level,
+  std::vector<int> parent_cell_indexes /* {-1} */)
+{
+  TimerOutput::Scope t(computing_timer, "finding_vertices_positions");
+
+  // Find the positions associated to the vertices of cell
+  for (const auto &cell : triangulation.cell_iterators_on_level(level))
+    {
+      bool parent_is_candidate = false;
+
+      // Check if parent cell of current level cell is a candidate
+      if (cell->level() > parameters.reconstruction_param.coarse_mesh_level)
```

Comment:
The coarse mesh is level is the coarsest level from which the search is started I guess?